src/NumericalAlgorithms/RootFinding/GslMultiRoot.hpp

```diff
@@ -0,0 +1,364 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+// \file
+// Defines function gsl_multiroot.
+
+#pragma once
+
+#include <array>
+#include <cmath>
+#include <cstddef>
+#include <exception>
+#include <gsl/gsl_errno.h>
+#include <gsl/gsl_matrix_double.h>
+#include <gsl/gsl_multiroots.h>
+#include <gsl/gsl_vector_double.h>
+#include <string>
+
+#include "ErrorHandling/Error.hpp"
+#include "Parallel/Printf.hpp"
+#include "Utilities/Gsl.hpp"
+
+namespace detail {
+template <size_t Dim, typename Solver>
+void print_state(size_t iteration_number, Solver* solver,
+                 bool print_header = false) noexcept {
+  if (print_header) {
+    Parallel::printf("Iter\t");
+    for (size_t i = 0; i < Dim; ++i) {
+      Parallel::printf(" x[%u]\t", i);
+    }
+    for (size_t i = 0; i < Dim; ++i) {
+      Parallel::printf(" f[%u]\t", i);
+    }
+    Parallel::printf("\n");
+  }
+
+  Parallel::printf("%u\t", iteration_number);
+  for (size_t i = 0; i < Dim; ++i) {
+    Parallel::printf("%3.4f  ", gsl_vector_get(solver->x, i));
+  }
+  for (size_t i = 0; i < Dim; ++i) {
+    Parallel::printf("%1.3e  ", gsl_vector_get(solver->f, i));
+  }
+  Parallel::printf("\n");
+}
+
+template <size_t Dim>
+std::array<double, Dim> gsl_to_std_array(const gsl_vector* const x) noexcept {
```

Comment:
Can the pointers to these functions be `gsl::not_null`s?

Reply:
From what I've tried, everything that expects a pointer to a gsl object must remain as it is

Reply:
Sounds good!